crates/holochain/src/core/workflow/app_validation_workflow/tests.rs

```diff
@@ -507,6 +507,119 @@ async fn validate_ops_in_sequence_must_get_action() {
+    let mut create = fixt!(Create);
+    create.entry_type = EntryType::App(AppEntryDef {
+        entry_index: 0.into(),
+        zome_index: 0.into(),
+        visibility: Default::default(),
+    });
+    let create_action = Action::Create(create);
+    let dht_create_op = DhtOp::RegisterAgentActivity(fixt!(Signature), create_action.clone());
+    let dht_create_op_hash = DhtOpHash::with_data_sync(&dht_create_op);
+    let dht_create_op_hashed = DhtOpHashed::from_content_sync(dht_create_op);
+
+    // create another op that will be validated successfully
+    let mut create = fixt!(Create);
+    create.entry_type = EntryType::App(AppEntryDef {
+        entry_index: 0.into(),
+        zome_index: 0.into(),
+        visibility: Default::default(),
+    });
+    let entry = fixt!(Entry);
+    let dht_store_entry_op =
+        DhtOp::StoreEntry(fixt!(Signature), NewEntryAction::Create(create), entry);
+    let dht_store_entry_op_hash = DhtOpHash::with_data_sync(&dht_store_entry_op);
+    let dht_store_entry_op_hashed = DhtOpHashed::from_content_sync(dht_store_entry_op);
+
+    // insert both ops in dht db and mark ready for app validation
+    app_validation_workspace.dht_db.test_write(move |txn| {
+        insert_op(txn, &dht_create_op_hashed).unwrap();
+        put_validation_limbo(txn, &dht_create_op_hash, ValidationStage::SysValidated).unwrap();
+        insert_op(txn, &dht_store_entry_op_hashed).unwrap();
+        put_validation_limbo(txn, &dht_store_entry_op_hash, ValidationStage::SysValidated).unwrap();
+    });
+
+    // check ops are now counted as ops to validate
+    let ops_to_validate =
+        validation_query::get_ops_to_app_validate(&app_validation_workspace.dht_db)
+            .await
+            .unwrap()
+            .len();
+    assert_eq!(ops_to_validate, 2);
+
+    let validation_dependencies = Arc::new(Mutex::new(ValidationDependencies::new()));
+
+    // running validation workflow should finish without errors
+    // outcome summary should show 1 validated and accepted op and the error-causing op as still to validate
+    let outcome_summary = app_validation_workflow_inner(
+        Arc::new(dna_hash.clone()),
+        app_validation_workspace.clone(),
+        conductor.raw_handle(),
+        &conductor.holochain_p2p().to_dna(dna_hash.clone(), None),
+        conductor
+            .get_or_create_space(&dna_hash)
+            .unwrap()
+            .dht_query_cache,
+        validation_dependencies.clone(),
+    )
+    .await
+    .unwrap();
+    assert_matches!(
+        outcome_summary,
+        OutcomeSummary {
+            ops_to_validate: 2,
+            validated: 1,
+            accepted: 1,
+            missing: 0,
+            rejected: 0
+        }
+    );
+
+    let ops_to_validate =
+        validation_query::get_ops_to_app_validate(&app_validation_workspace.dht_db)
+            .await
+            .unwrap()
+            .len();
+    assert_eq!(ops_to_validate, 1);
```

Comment:
So because of the error, this op has stayed ready to validate?

Can the retry logic be asserted here any any way? I guess that's above this level because we're calling the inner part of the workflow, so maybe that's just not relevant here?

Is it simple here to check which op didn't get validated? It'd be nice to tie the error to the op that was supposed to cause the error

Reply:
Yes, because of the error, the op is unvalidated as if it has been skipped. From my MM monolog:

> During app validation an op validation can return an error from the validate callback. That op is not considered validated, but also not considered as missing dependencies. Therefore app validation will not be re-triggered automatically and the op that couldn't be validated will be retried the next time some ops are coming in.

> Is that a problem? Looking at [the error that can be returned from a validate invocation](https://github.com/holochain/holochain/blob/develop/crates/holochain/src/core/workflow/app_validation_workflow/error.rs#L14), it seems that nearly all of them indicate a corrupt conductor state which cannot be recovered from easily, even the HolochainP2pErrors. After a conductor restart app validation will be run pretty soon, so I think it's not a problem.

Therefore retrying won't help, because it's nothing that the conductor state can recover from.

> Is it simple here to check which op didn't get validated? It'd be nice to tie the error to the op that was supposed to cause the error

It's simple, yes, although other than the test the hash wouldn't be used.

Reply:
Done